### PR TITLE
prairiedog: migrate to config file

### LINKED ISSUE #3622: OOTB: Remove model dependency from `prairiedog`
`prariedog` queries the API for all settings, deserializing into the model.  It should probably read from a template rendered by a settings extension.

### PR DESCRIPTION
**Pull request by jmt-lab**
**Issue number:** #3622 

Closes #3622 

**Description of changes:**
* Migrates prairiedog from fetching BootSettings from settings api to fetching this from /etc/prairiedog.toml or config file provided by -c/--config-path
* Adds template prairiedog-toml and migration

**Testing done:**
Fully tested migration upgrade and the reboot logic for the reconcile works. Also made sure kernel settings synced to the file

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: sources/api/prairiedog/src/bootconfig.rs [hunk 1]
```diff
@@ -94,23 +95,14 @@ where
...
-async fn get_boot_config_settings<P>(socket_path: P) -> Result<Option<BootSettings>>
+async fn get_boot_config_settings<P>(config_path: P) -> Result<Option<BootSettings>>
 where
     P: AsRef<Path>,
 {
-    let uri = "/settings";
-    let settings: serde_json::Value =
-        schnauzer::v1::get_json(socket_path, uri, Some(("prefix", "boot")))
-            .await
-            .context(error::RetrieveSettingsSnafu)?;
```
Comment (cbgbt) [inline sources/api/prairiedog/src/bootconfig.rs:105]: Mind removing the enum variants that aren't used anymore?

File: packages/os/prairiedog-toml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+[required-extensions]
+std = { version = "v1", helpers = ["default"]}
```
Comment (cbgbt) [inline packages/os/prairiedog-toml]: Looks like we don't actually use `default`. I think you can remove the dependency on `std`.

File: sources/api/migration/migrations/v1.18.1/prairiedog-config-file-v0-1-0/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+use migration_helpers::common_migrations::{
+    AddPrefixesMigration, ListReplacement, ReplaceListsMigration,
+};
+use migration_helpers::{migrate, Result};
+use std::process;
+
+fn run() -> Result<()> {
+    migrate(AddPrefixesMigration(vec![
+        "configuration-files.prairiedog-toml",
+    ]))?;
```
Comment (cbgbt) [inline sources/api/migration/migrations/v1.18.1/prairiedog-config-file-v0-1-0/src/main.rs]: Similar to the prairiedog PR, I don't think `AddPrefixesMigration` does the right thing, which means that I think the upgrade case will not actually begin rendering a new config file.

I think we need a new type of Migration (`AddConfigFileMigration`) which allows you to specify the values of the config file in the migration. On upgrade, the new migration is written as specified, and on downgrade it is removed.

**Review by cbgbt — COMMENTED** [review]
Cool! For some feedback on git style:
* I would squash `prairiedog: fix source number to match with other templates` to be included in the same commit that modified the RPM.
* We tend to leave PR/fixup style comments to the PRs themselves, and the commit messages just focus on the unique change made in that commit. For consistency, it would be best to remove the `fixup!` lines.

File: sources/api/prairiedog/src/bootconfig.rs (hunk 1 above)
Comment (jmt-lab) [reply sources/api/prairiedog/src/bootconfig.rs:105]: sure thing

File: sources/api/migration/migrations/v1.18.1/prairiedog-config-file-v0-1-0/src/main.rs (hunk 3 above)
Comment (jmt-lab) [reply sources/api/migration/migrations/v1.18.1/prairiedog-config-file-v0-1-0/src/main.rs]: This seems to work just fine from my testing

**Comment by jmt-lab** [comment]
squashed commits and updated changes via comments

File: sources/api/migration/migrations/v1.18.1/prairiedog-config-file-v0-1-0/src/main.rs (hunk 3 above)
Comment (webern) [reply sources/api/migration/migrations/v1.18.1/prairiedog-config-file-v0-1-0/src/main.rs]: Sean's comment concerns me. Can you do some additional testing?
- Use apiclient to change the relevant settings and confirm that the config file is re-rended with the changed settings.
- Confirm that the datastore is in the correct state after downgrade and the config file has been removed after downgrade.

File: packages/os/prairiedog-toml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+{{/each}}
+{{/if}}
+
+{{#if settings.boot.init}}
+[init]
+{{#each settings.boot.init}}
+"{{@key}}" = [ {{#each this}}"{{this}}",{{/each}} ]
+{{/each}}
+{{/if}}
+{{/if}}
```
Comment (bcressey) [inline packages/os/prairiedog-toml]: nit: can you structure templates like these so they render a completely empty file if there are no corresponding settings, and (ideally) with the correct spacing between the logical sections?

Whitespace issues like that are among my pet peeves when dealing with the output of templating systems. I'd like the output to match what I'd write if I created the file by hand, and as you may have noticed I have feelings about missing or extra newlines.

File: packages/os/prairiedog-toml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+{{#if settings.boot}}
+{{#if settings.boot.reboot-to-reconcile}}
+reboot-to-reconcile = {{settings.boot.reboot-to-reconcile}}
+{{/if}}
+
+{{#if settings.boot.kernel}}
+[kernel]
+{{#each settings.boot.kernel}}
+"{{@key}}" = [ {{#each this}}"{{this}}",{{/each}} ]
+{{/each}}
```
Comment (bcressey) [inline packages/os/prairiedog-toml]: Does this render correctly when the values contain quotes or spaces?
```
$ apiclient apply <<EOF
settings.boot.kernel-parameters]
param = ["\"with quotes\""]
EOF
```

File: sources/api/prairiedog/src/main.rs [hunk 6]
```diff
@@ -60,9 +62,9 @@ struct Args {
     #[argh(option, default = "LevelFilter::Info", short = 'l')]
     /// log-level trace|debug|info|warn|error
     log_level: LevelFilter,
-    #[argh(option, default = "constants::API_SOCKET.to_string()", short = 's')]
-    /// socket-path path to apiserver socket
-    socket_path: String,
+    #[argh(option, default = "DEFAULT_CONFIG_FILE.to_string()", short = 'c')]
+    /// config-path to provide alternate config file path
+    config_path: String,
```
Comment (bcressey) [inline sources/api/prairiedog/src/main.rs]: same question as elsewhere: can we use the `PathBuf` type for this?

File: sources/api/prairiedog/src/bootconfig.rs [hunk 7]
```diff
@@ -94,23 +95,14 @@ where
...
-        None => Ok(None),
-        Some(boot_settings_val) => Ok(Some(
-            serde_json::from_value(boot_settings_val.to_owned())
-                .context(error::BootSettingsFromJsonValueSnafu)?,
-        )),
-    }
+    let config_str = fs::read_to_string(config_path.as_ref()).context(error::ReadFileSnafu {
+        path: config_path.as_ref().to_path_buf(),
+    })?;
+    toml::from_str(config_str.as_str()).context(error::InputTomlSnafu)
```
Comment (bcressey) [inline sources/api/prairiedog/src/bootconfig.rs:123]: If we're not using await, then the function probably shouldn't be async and we maybe shouldn't even use tokio. On the other hand, if we're using async, we should use `tokio::fs` and await.

Right now this sort of splits the difference and ends up making blocking I/O calls in async functions, which is something we need to be vigilant about.

File: sources/api/migration/migrations/v1.18.1/prairiedog-config-file-v0-1-0/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    migrate(ReplaceListsMigration(vec![ListReplacement {
```
Comment (webern) [inline sources/api/migration/migrations/v1.18.1/prairiedog-config-file-v0-1-0/src/main.rs]: I believe, per Ben in a different PR, we cannot have multiple migrations in a single migration binary.

**Review by webern — CHANGES_REQUESTED** [review]

File: sources/api/prairiedog/src/bootconfig.rs (hunk 7 above)
Comment (jmt-lab) [reply sources/api/prairiedog/src/bootconfig.rs:123]: Most of the methods pre-existing in this tool are written with tokio async but without apiserver they technically don't seem to need to be, so the question really here is, would you like me to de-async this or just update my fs to use tokio::fs?

File: sources/api/migration/migrations/v1.18.1/prairiedog-config-file-v0-1-0/src/main.rs (hunk 8 above)
Comment (jmt-lab) [reply sources/api/migration/migrations/v1.18.1/prairiedog-config-file-v0-1-0/src/main.rs]: Performed the first test (first bullet point) it does indeed update the config file accordingly with the changed settings.

Comment (jmt-lab) [reply sources/api/migration/migrations/v1.18.1/prairiedog-config-file-v0-1-0/src/main.rs]: Performed the upgrade/downgrade test, downgrade works as expected

**Comment by jmt-lab** [comment]
Rebased off latest
Split migrations

**Comment by jmt-lab** [comment]
Adjust spaces in template

File: packages/os/prairiedog-toml (hunk 5 above)
Comment (jmt-lab) [reply packages/os/prairiedog-toml]: Isn't this more of a question for schnauzer than necessarily the template file, or is there a helper method I need to be using for the above?

File: sources/api/prairiedog/src/bootconfig.rs (hunk 7 above)
Comment (jmt-lab) [reply sources/api/prairiedog/src/bootconfig.rs:123]: De-asyncified prairiedog

**Review by webern — CHANGES_REQUESTED** [review]
Some lints are failing.

Also, we do not want to bump Release.toml and Twoliter.toml versions as part of this PR. That should be done separately then these PRs should be rebased.

**Comment by webern** [comment]
Let's merge this first #3768 then rebase your PR to align on 1.19.2

File: packages/os/prairiedog-toml (hunk 5 above)
Comment (jmt-lab) [reply packages/os/prairiedog-toml]: Yeah i need to change {{this}} to {{{this}}} otherwise it html escapes

**Comment by jmt-lab** [comment]
Rebased off latest
Fixed nested quotes in template
Fixed formatting

File: sources/api/prairiedog/Cargo.toml [hunk 9]
```diff
@@ -20,7 +20,7 @@ signpost = { path = "../../updater/signpost", version = "0.1" }
 simplelog = "0.12"
 snafu = "0.7"
 serde_json = "1"
-tokio = { version = "~1.32", default-features = false, features = ["macros", "rt-multi-thread"] } # LTS
```
Comment (webern) [inline sources/api/prairiedog/Cargo.toml:23]: One of the primary purposes of this work is to remove the dependency on `models`. Can you please remove it?

**Review by webern — CHANGES_REQUESTED** [review]
Ben's comments are blocking as well as my request that you remove the dependency on `models`. For future reference, removing the dependency on `models` is the reason we are doing this. If we forgot to do that on the previously merged PR, please do a follow up. Thanks!

**Comment by jmt-lab** [comment]
Models dependency removed

File: sources/api/prairiedog/Cargo.toml [hunk 10]
```diff
@@ -15,12 +15,14 @@ constants = { path = "../../constants", version = "0.1" }
 log = "0.4"
 nix = "0.26"
 models =  { path = "../../models", version = "0.1" }
```
Comment (webern) [inline sources/api/prairiedog/Cargo.toml]: This is the one that needs to be removed.

```suggestion
```

**Review by webern — CHANGES_REQUESTED** [review]
Close, but it still has the models dependency.

File: sources/api/prairiedog/Cargo.toml (hunk 10 above)
Comment (jmt-lab) [reply sources/api/prairiedog/Cargo.toml]: Fixed

**Review by webern — APPROVED** [review]

**Comment by jmt-lab** [comment]
rebased off latest

File: packages/os/prairiedog-toml [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+reboot-to-reconcile = {{settings.boot.reboot-to-reconcile}}
+
+{{/if}}
+{{#if settings.boot.kernel}}
+[kernel]
+{{#each settings.boot.kernel}}
+"{{@key}}" = [ {{#each this}}"{{{this}}}",{{/each}} ]
+{{/each}}
+
+{{/if}}
```
Comment (bcressey) [inline packages/os/prairiedog-toml]: Does this render a config file with two empty lines if none of the settings are present?

**Review by bcressey — APPROVED** [review]

**Pull request merged**